ckanext/s3filestore/uploader.py

```diff
@@ -427,21 +480,43 @@ def download(self, id, filename=None):
         except ClientError as ex:
             if ex.response['Error']['Code'] in ['NoSuchKey', '404']:
                 # attempt fallback
-                # TODO: use ckan.lib.uploader:ResourceUpload to get download link
+                default_resource_upload = DefaultResourceUpload(self.resource)
+                return default_resource_upload.download(id, self.filename)
+            else:
+                # Controller will raise 404 for us
+                raise OSError(errno.ENOENT)
+
+    def metadata(self, id, filename=None):
+        if filename is None:
+            filename = os.path.basename(self.url)
+        key_path = self.get_path(id, filename)
+        key = filename
+
+        if key is None:
+            log.warning('Key \'{0}\' not found in bucket \'{1}\''
```

Comment:
Is this the right message in this situation? We haven't actually made a call to the bucket yet.

Reply:
hmm, unsure why this is here as if filename is empty then it would say
Key 'None' not found in bucker 'abucket'